### ISSUE TITLE
Proper Train/Test Splits

### ISSUE THREAD

**Issue by valentinwust** [gao-lab/GLUE#71]
More a comment than a question, but I think you may want to be more careful when splitting data into test and training sets, and keeping them properly separated.

- GLUE appears to switch which cells it considers test cells when going from fit to finetune, so the test loss it reports during fine-tuning isn't accurate.
- If I understand your [CLUE code for the NeurIPS 2021 competition](https://github.com/openproblems-bio/neurips2021_multimodal_topmethods/blob/main/src/match_modality/methods/clue/run/script.py) correctly, you fine-tune on the test data before evaluating the pretrained model; but then it isn't really test data anymore. And even though you mask the test cell correspondences, given that you train on paired data I would say the discriminator setup still leaks some pair information that wouldn't be available in a realistic case of transferring knowledge from paired training data to unpaired test data. From the playing around with the model that I have done it doesn't look like this has a huge impact on the results, but test data should still not be used for any training. Best practice would of course be to hold out some validation data that is never used except for the final evaluation, but this should have been done by the competition organizers.

**Comment by XinmingTu** [comment]
Thanks a lot for your comment. However, I'm not entirely certain that I fully comprehend your feedback. 

So in a realistic scenario, we would have unpaired test data from various modalities, and we wouldn't have access to their corresponding information. Still, we can employ the unsupervised part of CLUE on this unpaired data. Could you possibly provide further clarification on your remarks?

**Comment by valentinwust** [comment]
First of all, test data is usually meant to be completely unseen by the model anyway, but let me clarify what I mean.

You first train on paired data with all correspondences known by the model, and afterwards with train and test data. But while you don't explicitly tell the model the corespondences in the test data, there nevertheless exists a one to one correspondence in the test data since it was obtained with a paired method. This is relevant since you use a discriminator to align the latent spaces of the different modalities (which as far as I can see is applied to both train and test data during finetuning?); it will try to obtain a perfect overlap, which is the correct solution anyway for paired data. But for unpaired test data this would no longer be the case, so while you are not using the direct cell correspondences, you are still implicitly using the fact that there is such a correspondence.

**Comment by XinmingTu** [comment]
I agree with your statement about the test data for supervised model, but for unsupervised/semi-supervised/self-supervised model, I think the model can use the test data without labels. 

Yes, totally, the discriminatory will try to align them, that's the design intuition for this unsupervised component. So even there is not a perfect correspondence in the test data, the adversarial learning loss can still align them.
